File: VillagesTown/Engines/PopulationEngine.swift
//
//  PopulationEngine.swift
//  VillagesTown
//
//  Created by Claude Code
//

import Foundation

class PopulationEngine {

    // MARK: - Population Growth
    func processPopulationGrowth(for villages: inout [Village]) {
        for i in 0..<villages.count {
            var village = villages[i]

            // Check food availability (1 food per 20 population - reduced requirement)
            let foodNeeded = max(1, village.population / 20)
            let currentFood = village.resources[.food] ?? 0

            if currentFood >= foodNeeded {
                // Enough food - population grows
                growPopulation(village: &village)
                // Consume food
                village.substract(.food, amount: foodNeeded)
            } else {
                // Starvation - population decreases
                starvePopulation(village: &village)
                // Consume all available food
                village.substract(.food, amount: currentFood)
            }

            villages[i] = village
        }
    }

    private func growPopulation(village: inout Village) {
        // MUCH higher base growth: 5% + flat bonus
        var growthRate = 0.05
        var flatBonus = 3  // Always grow by at least 3

        // Happiness bonus
        let happiness = village.totalHappiness
        if happiness >= 80 {
            growthRate += 0.03  // +3% if very happy
            flatBonus += 2
        } else if happiness >= 60 {
            growthRate += 0.01  // +1% if happy
            flatBonus += 1
        } else if happiness < 40 {
            growthRate -= 0.02  // -2% if unhappy
            flatBonus = 1
        }

        // Farm bonus - each farm adds growth
        let farmCount = village.buildings.filter { $0.name == "Farm" }.count
        flatBonus += farmCount * 2

        // Granary bonus
        if village.buildings.contains(where: { $0.name == "Granary" }) {
            growthRate += 0.02
        }

        // Check if at capacity
        if village.population >= village.populationCapacity {
            growthRate = 0
            flatBonus = 0
        }

        let percentGrowth = Int(Double(village.population) * growthRate)
        let totalGrowth = max(0, percentGrowth + flatBonus)

        if totalGrowth > 0 {
            village.modifyPopulation(by: totalGrowth)
            print("📈 \(village.name): +\(totalGrowth) pop (now \(village.population))")
        }
    }

    private func starvePopulation(village: inout Village) {
        // Starvation causes 5% population loss (reduced from 10%)
        let loss = max(1, Int(Double(village.population) * 0.05))
        village.modifyPopulation(by: -loss)

        // Happiness penalty
        village.modifyHappiness(by: -15)

        print("💀 \(village.name): STARVATION! Lost \(loss) population")
    }

    // MARK: - Happiness Management
    func processHappiness(for villages: inout [Village]) {
        for i in 0..<villages.count {
            var village = villages[i]

            // Start with decent base happiness
            var baseHappiness = 60

            // Food situation
            let foodNeeded = max(1, village.population / 20)
            let currentFood = village.resources[.food] ?? 0

            if currentFood >= foodNeeded * 2 {
                baseHappiness += 10  // Surplus food = happy
            } else if currentFood < foodNeeded {
                baseHappiness -= 25  // Starvation penalty
            }

            // Population density penalty if overcrowded
            if village.population > village.populationCapacity * 80 / 100 {
                baseHappiness -= 10
            }

            village.happiness = min(100, max(0, baseHappiness))
            villages[i] = village
        }
    }

    // MARK: - Tax Collection
    func collectTaxes(for villages: inout [Village]) {
        for i in 0..<villages.count {
            var village = villages[i]

            // Better tax: 1 gold per population
            let taxRate = 1.0
            let taxIncome = Int(Double(village.population) * taxRate)

            // Market bonus
            if village.buildings.contains(where: { $0.name == "Market" }) {
                let bonus = taxIncome / 4  // +25% with market
                village.add(.gold, amount: taxIncome + bonus)
                print("💰 \(village.name): +\(taxIncome + bonus) gold (market bonus)")
            } else {
                village.add(.gold, amount: taxIncome)
                print("💰 \(village.name): +\(taxIncome) gold")
            }

            villages[i] = village
        }
    }
}

File: villages_town_flutter/lib/engines/population_engine.dart
import 'dart:math';
import '../data/models/resource.dart';
import '../data/models/village.dart';
import 'game_manager.dart';

class PopulationEngine {
  void processPopulationGrowth(List<Village> villages) {
    for (var i = 0; i < villages.length; i++) {
      var village = villages[i];
      if (village.owner == 'neutral') continue;

      // Check food availability
      final foodNeeded = village.population ~/ 10;
      final hasFood = village.getResource(Resource.food) >= foodNeeded;

      if (hasFood) {
        village.subtractResource(Resource.food, foodNeeded);

        // Base growth: 5% + flat bonus
        var growthRate = 0.05;
        var flatBonus = 3;

        // Happiness bonuses
        if (village.happiness >= 70) {
          growthRate += 0.03;
          flatBonus += 2;
        } else if (village.happiness < 40) {
          growthRate -= 0.02;
        }

        // Farm bonus
        final farmCount = village.buildings.where((b) => b.name == 'Farm').length;
        flatBonus += farmCount * 2;

        // Granary bonus
        if (village.buildings.any((b) => b.name == 'Granary')) {
          growthRate += 0.02;
        }

        final growth = (village.population * growthRate).round() + flatBonus;
        village.modifyPopulation(growth);
      } else {
        // Starvation
        final loss = (village.population * 0.05).round();
        village.modifyPopulation(-loss);
        village.modifyHappiness(-15);
      }

      villages[i] = village;
    }
  }

  void collectTaxes(List<Village> villages) {
    final game = GameManager.shared;

    for (var i = 0; i < villages.length; i++) {
      final village = villages[i];
      if (village.owner == 'neutral') continue;

      // 1 gold per population
      var tax = village.population;

      // Market bonus +25%
      if (village.buildings.any((b) => b.name == 'Market')) {
        tax = (tax * 1.25).round();
      }

      game.modifyGlobalResource(village.owner, Resource.gold, tax);
    }
  }

  void processHappiness(List<Village> villages) {
    for (var i = 0; i < villages.length; i++) {
      var village = villages[i];
      if (village.owner == 'neutral') continue;

      // Baseline happiness tends toward 60
      var targetHappiness = 60;

      // Food surplus increases target
      if (village.getResource(Resource.food) > village.population ~/ 5) {
        targetHappiness += 10;
      }

      // Starvation decreases it severely
      if (village.getResource(Resource.food) < village.population ~/ 10) {
        targetHappiness -= 25;
      }

      // Overcrowding
      if (village.population > village.populationCapacity * 0.9) {
        targetHappiness -= 10;
      }

      // Drift toward target
      if (village.happiness < targetHappiness) {
        village.modifyHappiness(min(5, targetHappiness - village.happiness));
      } else if (village.happiness > targetHappiness) {
        village.modifyHappiness(-min(5, village.happiness - targetHappiness));
      }

      villages[i] = village;
    }
  }
}
